p-ata/src/processor.rs

```diff
@@ -0,0 +1,595 @@
+/// - **Tests**: Uses curve25519-dalek to replicate on-chain behavior  
+/// - **Other builds**: Returns `false`
+#[inline(always)]
+pub(crate) fn is_off_curve(_address: &Pubkey) -> bool {
+    #[cfg(target_os = "solana")]
+    {
+        const ED25519_CURVE_ID: u64 = 0;
+
+        let point_addr = _address.as_ref().as_ptr();
+
+        // SAFETY: We're passing valid pointers to the syscall
+        // The syscall directly returns the validation result:
+        // - 0 means point is ON the curve (valid)
+        // - 1 means point is OFF the curve (invalid)
+        // - any other value means error
+        let syscall_result = unsafe {
+            sol_curve_validate_point(ED25519_CURVE_ID, point_addr, core::ptr::null_mut())
+        };
+
+        syscall_result == 1
+    }
+    #[cfg(all(not(target_os = "solana"), test))]
+    {
+        // Host build (tests, benches): replicate the on-chain `sol_curve_validate_point` logic
+        // using curve25519-dalek. A pubkey is "off-curve" if it cannot be decompressed into
+        // an Edwards point **or** it decomposes to a small-order point
+        // (matches Solana’s runtime rules).
+
+        use curve25519_dalek::edwards::CompressedEdwardsY;
+
+        let mut bytes = MaybeUninit::<[u8; 32]>::uninit();
+        let bytes_ptr = bytes.as_mut_ptr() as *mut u8;
+        // SAFETY: We initialize all 32 bytes before calling assume_init()
+        let bytes = unsafe {
+            core::ptr::copy_nonoverlapping(_address.as_ref().as_ptr(), bytes_ptr, 32);
+            bytes.assume_init()
+        };
+        let compressed = CompressedEdwardsY(bytes);
+
+        match compressed.decompress() {
+            None => true,                    // invalid encoding → off-curve
+            Some(pt) => pt.is_small_order(), // small-order = off-curve, otherwise on-curve
+        }
+    }
+    #[cfg(all(not(target_os = "solana"), not(test)))]
+    {
+        false
+    }
+}
+
+/// Validate an expected bump and ensure no better canonical bump exists.
+///
+/// Given an expected bump, this function verifies that no higher bump value produces
+/// a valid (off-curve) PDA address. This prevents creation of non-canonical ATAs
+/// by rejecting sub-optimal bump seeds.
+///
+/// ## Arguments
+///
+/// * `seeds` - The PDA derivation seeds (wallet, token_program, mint)
+/// * `program_id` - The program ID for PDA derivation  
+/// * `expected_bump` - The bump value provided by the caller
+///
+/// ## Returns
+///
+/// * `(Some(address), bump)` - If a better bump was found, returns the canonical address and bump
+/// * `(None, expected_bump)` - If no better bump exists, returns the original expected_bump
+///
+/// ## Security
+///
+/// This function is critical for preventing PDA canonicality attacks. It ensures
+/// that only the highest valid bump can be used, maintaining deterministic
+/// address derivation across all clients.
+#[inline(always)]
+pub(crate) fn ensure_no_better_canonical_address_and_bump(
```

Comment:
nit: this function confused me a bit. It could be the `ensure` part of the function name confused me, or how it was used... how about calling it `try_find_better_program_derived_address_and_bump` and returning `Option<(Pubkey, u8)>`? Then it'll be clear if something better was found